### ISSUE TITLE
HM tutorial failed with "port already in use"

### ISSUE THREAD

**Issue by jondequinor** [equinor/ert#2072]
**Describe the bug**
The HM tutorial failed with `ert_shared.port_handler.PortAlreadyInUseException: Port 51820 already in use.`

**To Reproduce**
Uncertain, but most likely a factor of

- concurrent hm tutorial runs
- ensemble evaluator

**Expected behavior**
The CI setup isn't completely synthetic. It's not hard to imagine multiple experiements on one RGS node. This should be more robust.

**Enviromment**
 - OS: RHEL7
 - ERT/Komodo Release: 2010.10.b5
 - Python version: 3.6
 - Remote/HPC execution involved: no, but jenkins

**Additional context**
Build #678 of HM tutorial

**Comment by sondreso** [comment]
Looking at the code I think this is something that could happen if multiple runs of ert are started on the same node __at the same time__.

The point in the code where the `EvaluatorServerConfig` is instantiated 
(which is also when the available port is found) is here: 

https://github.com/equinor/ert/blob/48a1c2af375ae16cc84be044d42b8c0041d960d8/ert_shared/cli/main.py#L59-L59

Quite a few things happens before this actually bound here:  (honorable mentions are creation of the runpath which can take a substantial amount of time) 

https://github.com/equinor/ert/blob/71359963da5a034ea2db32d68c3a140c17f59bd3/ert_shared/ensemble_evaluator/evaluator.py#L179-L179

If this is correct it is not something that is likely to happen for normal use cases; however, we should be more robust against it. My first idea would be that we could actually reserve the port by binding the socket when the server config instance is created. Maybe @oysteoh has more insight into this?

**Comment by oysteoh** [comment]
My first reaction was to run and hide, but i guess i need to provide you with some insight... 😂   The problem i faced if i remember correctly is that you can only use that port from the same thread which has invoked `bind` . And that was not always the case for us where we first "found" available ports by binding to one, and then launched new threads with this port as an argument which then again did a `bind` to this port.  But this is from the top of my head right now , i might need to dig a bit into the code and try to remember more correctly if needed.

**Comment by jondequinor** [comment]
Having gained some insight into multiprocessing from recent bugs, you might have to pass the socket as an argument to the `_evaluate` process explicitly. I know multiprocessing will duplicate fds in some cases, which might fail for sockets?

Or did you already consider this?

**Comment by jondequinor** [comment]
The hm tutorial runs locally, but is currently using the "production" range of ports. This is overly restrictive, and can be thought of (with sufficiently many concurrent running tutorials) as not having a port range at all.

So we should allow the runner of ERT to decide the range and the range should, for CI, be the entire high port range.

**Comment by sondreso** [comment]
Question is now how we best include the option in the komodo test pipeline. Since setting it for komodo version <= 2021.10.b6 will result in an error. Any ideas?

**Comment by sondreso** [comment]
Guess this is very related to https://github.com/equinor/scout/issues/386

**Comment by sondreso** [comment]
I think I will just add it always to the HM-tutorial. If we have to bugfix 2021.09 we can manually set it without the port range on Jenkins. It will involve a bit more manual work, but not much. Everything else should be solved properly by solving https://github.com/equinor/scout/issues/386

**Comment by BjarneHerland** [comment]
Apologies for poking into this uninvited, but Jenkins yelled at me for failures this morning, which I eventually tracked to this topic.

My first thought is to modify

https://github.com/equinor/ert/blob/71359963da5a034ea2db32d68c3a140c17f59bd3/ert_shared/port_handler.py#L44-L45

to use the same loop/logic as

https://github.com/equinor/ert/blob/71359963da5a034ea2db32d68c3a140c17f59bd3/ert_shared/port_handler.py#L18-L21

and return (socket,portnum) instead of (host,portnum). Method-signature would have to be aligned and field _port_  should be updated (i.e. it may end up different from the value set here)

https://github.com/equinor/ert/blob/71359963da5a034ea2db32d68c3a140c17f59bd3/ert_shared/ensemble_evaluator/config.py#L116

This may clearly have side-effects (because the _port_-field in EvaluatorServerConfig is not correct before the EnsambleEvaluator runs), but it seems like a viable approach to me, FWIW.

**Comment by sondreso** [comment]
Yes, this was also our first idea! I think @oysteoh tried something similar to this when this was first implemented, but ran into some issues that was mentioned above. To quickly resolve the issue with the crashing HM tutorial, we decided to go with the solution of allowing the port range to be specified, which I think we should allow no matter if the above is implemented. However, we should also look into making things more robust, so any PRs with a solution along the lines sketch above are very welcome!

**Comment by BjarneHerland** [comment]
Being a sucker for challenges, I did a quick PoC for the suggested approach...

The problematic side-effect is that the port identified at this point

https://github.com/equinor/ert/blob/71359963da5a034ea2db32d68c3a140c17f59bd3/ert_shared/ensemble_evaluator/config.py#L116

is used to construct url/uris which are picked up and used by other system-compontents **before**  the EnsambleEvaluator executes. I.e (as hinted in comments above) identifying the port happens when instantiating EvaluatorServerConfig whereas binding to the port is done when EnsambleEvaluator executes, quite a while later.

To make the suggested strategy work, we need to ensure that 1) the port-value is updated properly in the EvaluatorServerConfig-instance, and 2) other system-components do not cache or try to use the port-value/uri before the EnsambleEvaluator executes.

I don't have sufficient overview of ERT, but this does seem possible. Can fiddle around with this a bit more next week if nothing else urgent comes along...

**Comment by jondequinor** [comment]
> 2) other system-components do not cache or try to use the port-value/uri before the EnsambleEvaluator executes.

Is that possible with high ports?

**Comment by BjarneHerland** [comment]
> > 1. other system-components do not cache or try to use the port-value/uri before the EnsambleEvaluator executes.
> 
> Is that possible with high ports?

Sorry - not following... My understanding is that the EnsambleEvaluator runs a server (websockets, apparently) and then the port we're discussing becomes part of urls (3, actually) identifying the server-instance. I assume those urls are used by other components to talk to the EnsambleEvaluator, and AFAIK there are no restrictions on the port-number except for the "normal ones". See e.g. the spec for WebSocket protocol

https://datatracker.ietf.org/doc/html/rfc6455#section-3

(Keeping in mind that I have onlyworked with ERT for a week and may not have a good understanding of its architecture...)

**Comment by jondequinor** [comment]
But the core issue is that (I believe):

1. **ERT instance 1** sees that port _n_ is available, saves _n_ to a config (the 3 URIs) and closes the socket.
2. **ERT instance 2** sees that the same _n_ port is available, saves _n_ to a config (the 3 URIs) and closes the socket.
3. The race is on…

So when you say 

> other system-components do not cache or try to use the port-value/uri before the EnsambleEvaluator executes.

you have to incude _other ERT instances_ as well. And I am not sure _how_ we're supposed to do that—except hold on to the socket, i.e. not close it.

**Comment by BjarneHerland** [comment]
Completely agree - classic race-condition. :smile: (Pretty close to a defining example, in fact)

And the solution is (as you also indicate) to hold on to the socket once you have bound it. I.e. avoid a time-gap between checking availability and aquiring.

The complicating factor seems to be this

> The problem i faced if i remember correctly is that you can only use that port from the same thread which has invoked `bind` . And that was not always the case for us where we first "found" available ports by binding to one, and then launched new threads with this port

Which, if correct, means we cannot bind in some main-thread and pass the socket itself to other threads. I am not entirely certain that this is true, however. Will check it out...

Also, to clarify:

> So when you say
> 
> > other system-components do not cache or try to use the port-value/uri before the EnsambleEvaluator executes.
> 
> you have to incude _other ERT instances_ as well. And I am not sure _how_ we're supposed to do that—except hold on to the socket, i.e. not close it.

I assume ERT instances only talk to "their own" EnsambleEvaluator. So the only components needing this URL are the ones within the same ERT instance...?

**Comment by jondequinor** [comment]
> Which, if correct, means we cannot bind in some main-thread and pass the socket itself to other threads. I am not entirely certain that this is true, however. Will check it out...

Yes, do that. I think however the problem lies with the `forkserver` we are running, essentially forcing a long-lived `socked` to pass from one process to the other. I'm not entirely sure how to do that.

**Comment by sondreso** [comment]
> I assume ERT instances only talk to "their own" EnsambleEvaluator. So the only components needing this URL are the ones within the same ERT instance...?

Yes, for now only components inside the same instance will need this information. In a slightly longer perspective we want to run this evaluator as a service, similar to how storage works today. But that will require quite a bit of refactoring anyways...

**Comment by BjarneHerland** [comment]
> Yes, do that. I think however the problem lies with the `forkserver` we are running, essentially forcing a long-lived `socked` to pass from one process to the other. I'm not entirely sure how to do that.

Mmm.... are we talking subprocesses (as opposed to threads)? If we are, I can imagine this is more complicated. But I'll dig into it.

**Comment by jondequinor** [comment]
> Mmm.... are we talking subprocesses (as opposed to threads)? If we are, I can imagine this is more complicated. But I'll dig into it.

no, we're talking `fork()` then copying of FDs as well as pickling any argument to `multiprocessing.Process`. There may be other things that are pickled, hard to say.

**Comment by sondreso** [comment]
Here is the place where multiprocessing is used: 

https://github.com/equinor/ert/blob/9c2905ef4b9abb8fd1700febae32504a1f037da8/ert_shared/ensemble_evaluator/ensemble/prefect.py#L197-L208

The config argument that is passed to this function is the same as we are discussing here. But we don't actually need the socket here, just the urls. So we could probably just save those, instead of passing the entire object (with the fd, if we chose to store the socket in the config object).

**Comment by BjarneHerland** [comment]
This seems to be the place where "other components" of an ERT instance are started and only three fields from  ee_config are used:  _dispatch_uri_  , _token_ and _cert_  . We can easily limit ourselves to pass only this information to the subprocess. In fact, the config-object will be  pickled and the client gets a copy of it, hence if the main-process changes the original config-object it will not be reflected in the client in any case.

It follows that we cannot rely on changing the port-value in the original config object like indicated in https://github.com/equinor/ert/issues/2072#issuecomment-926710103 **after clients are dispatched**.
